From c680c42bd8810b7b826be073caa0910c53eec049 Mon Sep 17 00:00:00 2001
From: Tanya Singh <tanya.singh@4ipnet.com>
Date: Wed, 11 Sep 2024 18:46:59 +0800
Subject: [PATCH 26/55] ugps: Add AFC related fields to 'ubus call gps info'
 command

Adds GST (error ellipse) and GSA (VDOP) NMEA sentence parsing to provide AFC-required uncertainty metrics for 6 GHz operation.

Signed-off-by: John Crispin <john@phrozen.org>
---
 .../patches/001-ubus-gps-add-gst-gsa.patch    | 78 +++++++++++++++++++
 1 file changed, 78 insertions(+)
 create mode 100644 package/utils/ugps/patches/001-ubus-gps-add-gst-gsa.patch

diff --git a/package/utils/ugps/patches/001-ubus-gps-add-gst-gsa.patch b/package/utils/ugps/patches/001-ubus-gps-add-gst-gsa.patch
new file mode 100644
index 0000000000..0333604813
--- /dev/null
+++ b/package/utils/ugps/patches/001-ubus-gps-add-gst-gsa.patch
@@ -0,0 +1,78 @@
+--- a/nmea.h
++++ b/nmea.h
+@@ -23,7 +23,7 @@
+ 
+ #include <libubox/ustream.h>
+ 
+-extern char longitude[33], latitude[33], course[17], speed[17], elevation[17], satellites[3], hdop[5];
++extern char longitude[33], latitude[33], course[17], speed[17], elevation[17], satellites[3], hdop[5], major_axis[17], minor_axis[17], major_orientation[17], vdop[17];
+ extern int nmea_open(char *dev, struct ustream_fd *s, speed_t speed);
+ extern void gps_timestamp(void);
+ extern unsigned int adjust_clock;
+@@ -35,5 +35,8 @@ extern char gps_fields;
+ #define GPS_FIELD_ALT (1<<4)
+ #define GPS_FIELD_SAT (1<<5)
+ #define GPS_FIELD_HDP (1<<6)
++#define GPS_FIELD_MAJ (1<<7)
++#define GPS_FIELD_MIN (1<<8)
++#define GPS_FIELD_VDOP (1<<9)
+ 
+ #endif
+--- a/nmea.c
++++ b/nmea.c
+@@ -52,7 +52,8 @@ struct nmea_param {
+ } nmea_params[MAX_NMEA_PARAM];
+ 
+ static int nmea_bad_time;
+-char longitude[33] = { 0 }, latitude[33] = { 0 }, course[17] = { 0 }, speed[17] = { 0 }, elevation[17] = { 0 }, satellites[3] = { 0 }, hdop[5] = { 0 };
++char longitude[33] = { 0 }, latitude[33] = { 0 }, course[17] = { 0 }, speed[17] = { 0 }, elevation[17] = { 0 }, satellites[3] = { 0 }, hdop[5] = { 0 },
++     major_axis[17] = { 0 }, minor_axis[17] = { 0 }, major_orientation[17] = { 0 }, vdop[17] = { 0 };
+ int gps_valid = 0;
+ char gps_fields = 0;
+ 
+@@ -243,6 +244,30 @@ nmea_vtg_cb(void)
+ 	DEBUG(4, "speed: %s\n", speed);
+ }
+ 
++static void
++nmea_gst_cb(void)
++{
++	if (!gps_valid)
++		return;
++	strncpy(major_axis, nmea_params[3].str, sizeof(major_axis));
++	strncpy(minor_axis, nmea_params[4].str, sizeof(minor_axis));
++	strncpy(major_orientation, nmea_params[5].str, sizeof(major_orientation));
++	gps_fields |= GPS_FIELD_MAJ | GPS_FIELD_MIN;
++	DEBUG(4, "major_axis: %s\n", major_axis);
++	DEBUG(4, "minor_axis: %s\n", minor_axis);
++	DEBUG(4, "major_orientation: %s\n", major_orientation);
++}
++
++static void
++nmea_gsa_cb(void)
++{
++	if (!gps_valid)
++		return;
++	strncpy(vdop, nmea_params[17].str, sizeof(vdop));
++	gps_fields |= GPS_FIELD_VDOP;
++	DEBUG(4, "vdop: %s\n", vdop);
++}
++
+ static struct nmea_msg {
+ 	char *msg;
+ 	int cnt;
+@@ -272,6 +297,14 @@ static struct nmea_msg {
+ 		.msg = "ZDA",
+ 		.cnt = 5,
+ 		.handler = nmea_zda_cb,
++	}, {
++		.msg = "GST",
++		.cnt = 8,
++		.handler = nmea_gst_cb,
++	}, {
++		.msg = "GSA",
++		.cnt = 18,
++		.handler = nmea_gsa_cb,
+ 	},
+ };
+ 
-- 
2.34.1

